src/types.rs

```diff
@@ -320,36 +318,41 @@ impl Type {
         }
     }
 
-    /// Substitute the specified [TypeArg]s for type variables in this type.
-    ///
-    /// # Arguments
-    ///
-    /// * `args`: values to substitute in; there must be at least enough for the
-    /// typevars in this type (partial substitution is not supported).
-    ///
-    /// * `extension_registry`: for looking up [TypeDef]s in order to recompute [TypeBound]s
-    /// as these may get narrower after substitution
-    ///
-    /// # Panics
-    ///
-    /// If a [TypeArg] (that is referenced by a typevar in this type) does not contain a [Type],
-    /// contains a type with an incorrect [TypeBound], or there are not enough `args`.
-    /// These conditions can be detected ahead of time by [Type::validate]ing against the [TypeParam]s
-    /// and [check_type_args]ing the [TypeArg]s against the [TypeParam]s.
-    pub(crate) fn substitute(&self, sub: &Substitution) -> Self {
+    pub(crate) fn transform(&self, t: &impl TypeTransformer) -> Self {
         match &self.0 {
             TypeEnum::Prim(PrimType::Alias(_)) | TypeEnum::Sum(SumType::Simple { .. }) => {
                 self.clone()
             }
-            TypeEnum::Prim(PrimType::Variable(idx, bound)) => sub.apply_to_type_var(*idx, *bound),
-            TypeEnum::Prim(PrimType::Extension(cty)) => Type::new_extension(cty.substitute(sub)),
-            TypeEnum::Prim(PrimType::Function(bf)) => Type::new_function(bf.substitute(sub)),
-            TypeEnum::Tuple(elems) => Type::new_tuple(sub.apply_row(elems)),
-            TypeEnum::Sum(SumType::General { row }) => Type::new_sum(sub.apply_row(row)),
+            TypeEnum::Prim(PrimType::Variable(idx, bound)) => t.apply_typevar(*idx, *bound),
+            TypeEnum::Prim(PrimType::Extension(cty)) => Type::new_extension(cty.transform(t)),
+            TypeEnum::Prim(PrimType::Function(bf)) => Type::new_function(bf.transform(t)),
+            TypeEnum::Tuple(elems) => Type::new_tuple(transform_row(elems, t)),
+            TypeEnum::Sum(SumType::General { row }) => Type::new_sum(transform_row(row, t)),
         }
     }
 }
 
+pub(crate) trait TypeTransformer {
+    fn apply_typevar(&self, idx: usize, bound: TypeBound) -> Type {
```

Comment:
I think a dedicated type/newtype for type var index would help disambiguate a bit

Reply:
Great idea, thanks, will do